patch-package
--- a/node_modules/interface-connection/src/connection.js
+++ b/node_modules/interface-connection/src/connection.js
@@ -1,17 +1,26 @@
 'use strict'
 
 const defer = require('pull-defer/duplex')
+const pull = require('pull-stream')
+const abortable = require('pull-abortable')
+
+const noop = () => {}
 
 module.exports = class Connection {
   constructor (conn, info) {
     this.peerInfo = null
     this.conn = defer()
+    this.resolved = false
+    this.aborter = abortable()
 
     if (conn) {
       this.setInnerConn(conn, info)
     } else if (info) {
       this.info = info
     }
+
+    pull(this, this.aborter)
+    pull(this.aborter, this)
   }
 
   get source () {
@@ -22,6 +31,16 @@ module.exports = class Connection {
     return this.conn.sink
   }
 
+  close (err, callback) {
+    if (typeof err === 'function') {
+      callback = err
+      err = undefined
+    }
+    callback = callback || noop
+    this.aborter.abort(err)
+    callback()
+  }
+
   getPeerInfo (callback) {
     if (this.info && this.info.getPeerInfo) {
       return this.info.getPeerInfo(callback)
@@ -51,6 +70,7 @@ module.exports = class Connection {
 
   setInnerConn (conn, info) {
     this.conn.resolve(conn)
+    this.resolved = true
     if (info) {
       this.info = info
     } else {
