patch-package
--- a/node_modules/libp2p-switch/src/stats/stat.js
+++ b/node_modules/libp2p-switch/src/stats/stat.js
@@ -160,12 +167,14 @@ class Stats extends EventEmitter {
    * @returns {void}
    */
   _updateFrequency (latestTime) {
-    const timeDiff = latestTime - this._frequencyLastTime
+    const timeDiff = (latestTime - this._frequencyLastTime) || 1
 
     Object.keys(this._stats).forEach((key) => {
       this._updateFrequencyFor(key, timeDiff, latestTime)
     })
 
     this._frequencyLastTime = latestTime
   }
 